sky/backends/cloud_vm_ray_backend.py

```diff
@@ -4091,31 +4033,41 @@ def post_teardown_cleanup(self,
         log_abs_path = os.path.abspath(log_path)
         cluster_name_on_cloud = handle.cluster_name_on_cloud
 
+        # Backward compatibility for TPU nodes created before #2943.
+        # Deletion of TPU is done in SkyPilot's new GCP provisioner API
+        # TODO (zhwu): Remove this after 0.6.0.
         if (handle.tpu_delete_script is not None and
                 os.path.exists(handle.tpu_delete_script)):
-            with rich_utils.safe_status('[bold cyan]Terminating TPU...'):
-                tpu_rc, tpu_stdout, tpu_stderr = log_lib.run_with_log(
-                    ['bash', handle.tpu_delete_script],
-                    log_abs_path,
-                    stream_logs=False,
-                    require_outputs=True)
-            if tpu_rc != 0:
-                if _TPU_NOT_FOUND_ERROR in tpu_stderr:
-                    logger.info('TPU not found. '
-                                'It should have been deleted already.')
-                elif purge:
-                    logger.warning(
-                        _TEARDOWN_PURGE_WARNING.format(
-                            reason='stopping/terminating TPU',
-                            details=tpu_stderr))
-                else:
-                    raise RuntimeError(
-                        _TEARDOWN_FAILURE_MESSAGE.format(
-                            extra_reason='It is caused by TPU failure.',
-                            cluster_name=common_utils.cluster_name_in_hint(
-                                handle.cluster_name, cluster_name_on_cloud),
-                            stdout=tpu_stdout,
-                            stderr=tpu_stderr))
+            # Only call the deletion script if the cluster config does not
+            # contain TPU node config. Otherwise, the deletion should
+            # already be handled by the new provisioner.
```

Comment:
(1) Is the following correct:
- TPU VMs: launch/stop/delete is handled entirely in provisioner v2
- TPU Nodes: launch/stop/delete is handled entirely in provisioner v2

(2) I'm a bit confused by the code block here. When will L4039-4040 and L4046 be true? For old or new TPU nodes?

Reply:
Sg, moved this comment into a suggestion.